libraries/Microsoft.Bot.Connector/Authentication/BuiltinCloudEnvironment.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Logging;
+using Microsoft.Rest;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    internal abstract class BuiltinCloudEnvironment : ICloudEnvironment
+    {
+        private string _toChannelFromBotOAuthScope;
+        private string _loginEndpoint;
+        private string _callerId;
+        private string _channelService;
+        private IServiceClientCredentialsFactory _credentialFactory;
+        private AuthenticationConfiguration _authConfiguration;
+        private HttpClient _httpClient;
+        private ILogger _logger;
+
+        public BuiltinCloudEnvironment(string toChannelFromBotOAuthScope, string loginEndpoint, string callerId, string channelService, IServiceClientCredentialsFactory credentialFactory, AuthenticationConfiguration authConfiguration, HttpClient httpClient, ILogger logger)
+        {
+            _toChannelFromBotOAuthScope = toChannelFromBotOAuthScope;
```

Comment:
Seems like basic validations would be useful here? Proposing them for your convenience but let me know if I'm missing somehting and they are not needed.

```suggestion
            if (string.IsNullOrEmpty(toChannelFromBotOAuthScope))
            {
                throw new ArgumentNullException(nameof(toChannelFromBotOAuthScope));
            }
            
            if (string.IsNullOrEmpty(loginEndpoint))
            {
                throw new ArgumentNullException(nameof(loginEndpoint));
            }
            
            if (string.IsNullOrEmpty(callerId))
            {
                throw new ArgumentNullException(nameof(callerId));
            }

            if (string.IsNullOrEmpty(channelService))
            {
                throw new ArgumentNullException(nameof(channelService));
            }
            
            _toChannelFromBotOAuthScope = toChannelFromBotOAuthScope;
```

Reply:
Maybe true - this is internal however - and possible temporary - ideally I whack these classes and just end out with the "Parameterized" one.

Reply:
Well when I say internal i really mean from config I guess. I'll evaluate this and see how well we can protect ourselves from incorrect configuration. The internal code paths are dictated by the ChannelService value - its its not recognized then the remained of the fields are ignored.